Fix for #449 - Enhanced tree map visualization

fo-code opened the pull request:
In order to make the tree map easier to read, I changed the following:

- the tree map is shown on the whole screen now 
- the border color saturation is alternating, following the [echarts documentation](https://echarts.apache.org/en/option.html#series-treemap.itemStyle.borderColorSaturation), which says that an alternating high color saturation is a good option for distinguishing different sections.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/model/CoverageViewModel/index.jelly:
```diff
@@ -3,7 +3,7 @@
   <st:header name="Content-Type" value="text/html;charset=UTF-8"/>
 
-  <bs:page it="${it}" class="fluid-container d-flex flex-column">
+  <bs:page it="${it}" class="fluid-container d-flex flex-column h-100">
```
This breaks the visualization on smaller displays.

uhafner replied:
Can we remove these unrelated changes? Then we can merge the color PR and create a new release.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/model/CoverageViewModel/index.jelly:
```diff
@@ -42,35 +42,34 @@
     </ul>
 
-    <div class="tab-content">
+    <div class="tab-content h-100">
```
See above

uhafner commented on plugin/src/main/webapp/js/charts.js:
```diff
@@ -168,63 +168,63 @@ const CoverageChartGenerator = function ($) {
-                        gapWidth: 2,
-                        borderColorSaturation: 0.7
+                        gapWidth: 3,
+                        borderColorSaturation: 0.8
```
In total I think the greens look better, but the reds not.

uhafner reviewed (approved):
Would it help to use white as text color?

uhafner commented:
> the tree map is shown on the whole screen now
The problem with that approach is, that you then need to scroll on small displays. I found no way yet to always use the whole visible screen only for the trees. Maybe the problem is that the side panel is too large.

uhafner reviewed (changes requested):
> the tree map is shown on the whole screen now

The problem with that approach is, that you then need to scroll on small displays. I found no way yet to always use the whole visible screen only for the trees. Maybe the problem is that the side panel is too large.

fo-code commented:
> The problem with that approach is, that you then need to scroll on small displays. I found no way yet to always use the whole visible screen only for the trees. Maybe the problem is that the side panel is too large.

Okay, well I do not have a small screen and when I try it by resizing the window, it fills just the visible screen due to the resize event, but this might behave differently on a real small screen.
Nonetheless, I think in most cases `h-100` is useful, even if a few users have to scroll. 
I guess not that many developer use a small laptop screen all the time and even if it is the case, a bigger tree map allows that more file names are displays and the scrolling is limited.
What do you think about that?

> In total I think the greens look better, but the reds not.

I totally agree - I tried some different approaches and found a better solution. 
Also, I recognized that the package colors are generated by the echarts API, but these should be set dependend on the coverage.
Furthermore, the alpha of the currently used colors are problematic in the case of the map, so I removed these.
Therefore, I locally changed some values including the echarts plugin and the result can be shown above - I updated the "after" examples.
The new approach shows the coverage of each node exactly and different levels are clearly separable.

I am not available for two weeks now, so I would update the code afterwards - but maybe you can already give me some feedback on the new approach?

uhafner commented:
> > The problem with that approach is, that you then need to scroll on small displays. I found no way yet to always use the whole visible screen only for the trees. Maybe the problem is that the side panel is too large.
> 
> Okay, well I do not have a small screen and when I try it by resizing the window, it fills just the visible screen due to the resize event, but this might behave differently on a real small screen. 
> Nonetheless, I think in most cases `h-100` is useful, even if a few users have to scroll. I guess not that many developer use a small laptop screen all the time and even if it is the case, a bigger tree map allows that more file names are displays and the scrolling is limited. What do you think about that?

I would rather prefer to create different views for small and large screens. This shouldn't be too hard with Bootstraps classes.  See https://github.com/jenkinsci/code-coverage-api-plugin/blob/master/plugin/src/main/resources/coverage/coverage-table.jelly#L14 for an example

(See https://getbootstrap.com/docs/5.2/layout/grid/ for a documentation of all responsive classes)

> 
> > In total I think the greens look better, but the reds not.
> 
> I totally agree - I tried some different approaches and found a better solution. Also, I recognized that the package colors are generated by the echarts API, but these should be set dependend on the coverage. Furthermore, the alpha of the currently used colors are problematic in the case of the map, so I removed these. Therefore, I locally changed some values including the echarts plugin and the result can be shown above - I updated the "after" examples. The new approach shows the coverage of each node exactly and different levels are clearly separable.
> 
> I am not available for two weeks now, so I would update the code afterwards - but maybe you can already give me some feedback on the new approach?

The new colors look better now. I wonder if we should use the red, orange and green colors of Jenkins as well here? 
https://weekly.ci.jenkins.io/design-library/Colors/

fo-code commented:
> I would rather prefer to create different views for small and large screens.

So you mean we can add `h-100` for large screens only in order to use the full screen here but also support small screens without `h-100`?


> I wonder if we should use the red, orange and green colors of Jenkins as well here?

Yes of course, so you mean to change the currently used colors in `CoverageColorPalette` and insert the RGB codes of the Jenkins colors manually? Or is there a possibility to load the library colors dynamically as for example RGB values in Java?

uhafner commented:
> > I would rather prefer to create different views for small and large screens.
> 
> So you mean we can add `h-100` for large screens only in order to use the full screen here but also support small screens without `h-100`?
 
Yes, that shouldn't be too hard to implement.

> > I wonder if we should use the red, orange and green colors of Jenkins as well here?
> 
> Yes of course, so you mean to change the currently used colors in `CoverageColorPalette` and insert the RGB codes of the Jenkins colors manually? Or is there a possibility to load the library colors dynamically as for example RGB values in Java?

Getting them dynamically should work for the charts if we change the proxy methods:

So rather than using
```
            viewProxy.getOverview(function (t) {
                createOverview(t.responseObject(), 'coverage-overview');
            });
```
we could use

```
            viewProxy.getOverview(colors, function (t) {
                createOverview(t.responseObject(), 'coverage-overview');
            });
```

and fill `colors` with a JSON object, that contains the important styles. Something like:
```
{
 green: '#00ff00',
 orange: '#00ff00',
 red: '#00ff00'
}
```
(See https://weekly.ci.jenkins.io/design-library/Colors/ for the available colors).

You can get the colors then via:
```
    function getTextColor() {
        return getComputedStyle(document.body).getPropertyValue('--text-color') || '#333';
    }
```

Then the colors automatically will be adapted to the selected theme.

fo-code commented:
> Yes, that shouldn't be too hard to implement.

I agree, this looks like a good solution.


> Getting them dynamically should work for the charts if we change the proxy methods:

Thanks for the explanation.
As I understand this, the color JSON has to be set manually? In this case, I prefer using an Enum since this allows using these colors within the whole plugin and it is easier to integrate.
More specific what I wanted to know is if for example this color JSON can be requested or if I can call a Java API to request these color hex codes within the Java code of the coverage plugin? I am not really familiar of how the design library works...

uhafner commented:
> > Yes, that shouldn't be too hard to implement.
> 
> I agree, this looks like a good solution.
> 
> > Getting them dynamically should work for the charts if we change the proxy methods:
> 
> Thanks for the explanation. As I understand this, the color JSON has to be set manually? In this case, I prefer using an Enum since this allows using these colors within the whole plugin and it is easier to integrate. More specific what I wanted to know is if for example this color JSON can be requested or if I can call a Java API to request these color hex codes within the Java code of the coverage plugin? I am not really familiar of how the design library works...

It works the other was round: there is no way for the server to call the client (and there is no API on the client side). You need to wrap all information in an existing Ajax call from the client to the server:

In Jenkins processing is started on the client where the user selects the HTTP address. This is mapped on the server side into a jelly view. This jelly view is then rendered as HTML and send to the client that will render it. In this view we can invoke an associated JS file and call back to the server using Ajax. I.e., when the client currently starts a request to get the coverage data (the chart is empty in the jelly view) the client can provide an additional parameter (that is the reason that I am proposing a JSON object on the JS side, where we can put everything into what we want. I use the same thing in the trend chart configuration). This JSON can be interpreted on the server side to extract the colors. An enum will not work as it is constant on the server side, but actually each client might have its own colors. Maybe we can setup a call to discuss this part in more detail?

fo-code commented:
Yes, this would be perfect! I sent you an email.

fo-code commented:
@uhafner I finished the color rework.
The Jenkins colors from the Design Library are now in use and for the table and the tree map dynamically loaded.
The fallback colors correspond to the Jenkins colors aswell.
Currently, some HSL colors are provided in the dark mode - these cases are not handled yet and the fallback is used.
The size of the tree map is like before in order to work on all screen sizes.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/model/CoverageTableModel.java:
```diff
@@ -206,12 +215,12 @@ public int getLoc() {
                                         .withStyle(String.format(
                                                 "background-image: linear-gradient(90deg, %s %f%%, transparent %f%%);",
-                                                colors.getFillColorAsHex(),
+                                                colors.getFillColorAsRGBAHex(80),
```
80 seems to be some kind of random, maybe it helps to define a constant with this value that uses a appropriate name?

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/model/visualization/colorization/CoverageColorJenkinsId.java:
```diff
@@ -0,0 +1,45 @@
+
+    private final String jenkinsColorId;
+
+    /**
+     * Constructor.
+     *
+     * @param colorId
+     *         The color id
+     */
```
This can be deleted

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/model/CoverageViewModel/index.jelly:
```diff
@@ -44,35 +45,34 @@
     </div>
 
-    <div class="tab-content">
+    <div class="tab-content pt-2">
```
See above

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/model/CoverageViewModel/index.jelly:
```diff
@@ -44,35 +45,34 @@
+    <div class="tab-content pt-2">
 
-      <div role="tabpanel" id="overview" class="tab-pane fade" aria-labelledby="overview-tab" >
+      <div role="tabpanel" id="overview" class="tab-pane fade h-100" aria-labelledby="overview-tab" >
```
See above

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/model/CoverageViewModel/index.jelly:
```diff
@@ -44,35 +45,34 @@
           <div class="col-5">
             <bs:card title="${%Total coverage overview}" fontAwesomeIcon="chart-bar">
-              <div id="coverage-overview" class="graph-cursor-pointer overview-chart h-100"/>
+              <div id="coverage-overview" class="graph-cursor-pointer overview-chart"/>
```
See above

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/model/CoverageViewModel/index.jelly:
```diff
@@ -44,35 +45,34 @@
           <div class="col-7">
             <bs:card title="${%Coverage trend}" fontAwesomeIcon="chart-line">
-              <div id="coverage-trend" class="graph-cursor-pointer overview-chart h-100"/>
+              <div id="coverage-trend" class="graph-cursor-pointer overview-chart"/>
```
See above

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/model/CoverageViewModel/index.jelly:
```diff
@@ -44,35 +45,34 @@
+      <div role="tabpanel" id="lineCoverage" class="tab-pane fade h-100" aria-labelledby="lineCoverage-tab">
+        <bs:card title="${%Line coverage}" fontAwesomeIcon="folder-tree"
+                 class="flex-fill h-100" bodyClass="d-flex flex-column h-100">
+          <div id="project-line-coverage" class="graph-cursor-pointer tree-chart flex-grow-1"/>
```
See above

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/model/CoverageViewModel/index.jelly:
```diff
@@ -44,35 +45,34 @@
+      <div role="tabpanel" id="branchCoverage" class="tab-pane fade h-100" aria-labelledby="branchCoverage-tab">
+        <bs:card title="${%Branch coverage}" fontAwesomeIcon="folder-tree"
+                 class="flex-fill h-100" bodyClass="d-flex flex-column h-100">
+          <div id="project-branch-coverage" class="graph-cursor-pointer tree-chart flex-grow-1"/>
```
See above

uhafner reviewed (changes requested):
I like the new design! The only mandatory changes are the viewpoint adjustments which are unrelated to this PR and should be changed in a different PR.

fo-code commented:
Nice to hear and thanks for the review!
Then, everything should be ready to be released. :)
